src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -0,0 +1,503 @@
+        internal ManualResetEvent EnsureExitedEvent()
+        {
+            Debug.Assert(!Monitor.IsEntered(_gate));
+
+            lock (_gate)
+            {
+                // If we already have an initialized event, just return it.
+                if (_exitedEvent == null)
+                {
+                    // If we don't, create one, and if the process hasn't yet exited,
+                    // make sure we have a task that's actively monitoring the completion state.
+                    _exitedEvent = new ManualResetEvent(initialState: _exited);
+                    if (!_exited)
+                    {
+                        // If we haven't exited, we need to spin up an asynchronous operation that
+                        // will completed the exitedEvent when the other process exits. If there's already
+                        // another operation underway, then we'll just tack ours onto the end of it.
+                        _waitInProgress = _waitInProgress == null ?
+                            WaitForExitAsync() :
+                            _waitInProgress.ContinueWith((_, state) => ((ProcessWaitState)state).WaitForExitAsync(),
+                                this, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default).Unwrap();
+                    }
+                }
+                return _exitedEvent;
+            }
+        }
+
+        internal DateTime ExitTime
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    Debug.Assert(_exited);
+                    return _exitTime;
+                }
+            }
+        }
+
+        internal bool HasExited
+        {
+            get
+            {
+                int? ignored;
+                return GetExited(out ignored);
+            }
+        }
+
+        internal bool GetExited(out int? exitCode)
+        {
+            lock (_gate)
+            {
+                // Have we already exited?  If so, return the cached results.
+                if (_exited)
+                {
+                    exitCode = _exitCode;
+                    return true;
+                }
+
+                // Is another wait operation in progress?  If so, then we haven't exited,
+                // and that task owns the right to call CheckForExit.
+                if (_waitInProgress != null)
+                {
+                    exitCode = null;
+                    return false;
+                }
+
+                // We don't know if we've exited, but no one else is currently
+                // checking, so check.
+                CheckForExit();
+
+                // We now have an up-to-date snapshot for whether we've exited,
+                // and if we have, what the exit code is (if we were able to find out).
+                exitCode = _exitCode;
+                return _exited;
+            }
+        }
+
+        private void CheckForExit(bool blockingAllowed = false)
```

Comment:
I'd like to leave this here as a reminder that it's really something we should do, or should at least investigate to see what kind of impact it has on perf (I expect it will be a valuable thing to do).  For now, I've added the suggested comments as well as a Debug.Assert that validates it's not set to true until we do more work.